components/service/glean/src/main/java/mozilla/components/service/glean/EventMetricType.kt

```diff
@@ -0,0 +1,107 @@
+import mozilla.components.support.base.log.logger.Logger
+
+// Maximum length of any string value in the extra dictionary, in UTF8 byte sequence length.
+const val MAX_LENGTH_EXTRA_KEY_VALUE = 80
+// Maximum length of any passed value string, in UTF8 byte sequence length.
+const val MAX_LENGTH_VALUE = 80
+// Maximum length of any passed object string, in UTF8 byte sequence length.
+const val MAX_LENGTH_OBJECT_ID = 20
+
+/**
+ * This implements the developer facing API for recording events.
+ *
+ * Instances of this class type are automatically generated by the parsers at built time,
+ * allowing developers to record events that were previously registered in the metrics.yaml file.
+ *
+ * The Events API only exposes the [record] method, which takes care of validating the input
+ * data and making sure that limits are enforced.
+ */
+data class EventMetricType(
+    override val applicationProperty: Boolean,
+    override val disabled: Boolean,
+    override val group: String,
+    override val name: String,
+    override val sendInPings: List<String>,
+    override val userProperty: Boolean,
+    val allowedExtraKeys: List<String>? = null
+) : CommonMetricData {
+    private val logger = Logger("glean/EventMetricType")
+
+    /**
+     * Record an event by using the information provided by the instance of this class.
+     *
+     * @param objectId the object the event occurred on, e.g. 'reload_button'. The maximum
+     *                 length of this string is defined by [MAX_LENGTH_OBJECT_ID]
+     * @param value optional. This is a user defined value, providing context for the event. The
+     *              maximum length of this string is defined by [MAX_LENGTH_VALUE]
+     * @param extra optional. This is map, both keys and values need to be strings, keys are
+     *              identifiers. This is used for events where additional richer context is needed.
+     *              The maximum length for values is defined by [MAX_LENGTH_EXTRA_KEY_VALUE]
+     */
+    public fun record(objectId: String, value: String? = null, extra: Map<String, String>? = null) {
+        // TODO: report errors through other special metrics handled by the SDK. See bug 1499761.
+
+        // Silently drop recording for disabled events.
+        if (disabled) {
+            return
+        }
+
+        // TODO: implement "user" metric lifetime. See bug 1499756.
+        // Metrics can be recorded with application or user lifetime. For now,
+        // we only support "application": metrics live as long as the application lives.
+        if (!applicationProperty) {
+            logger.error("The metric lifetime must be explicitly set.")
+            return
+        }
+
+        val truncatedObjectId = objectId.let {
+            if (it.length > MAX_LENGTH_OBJECT_ID) {
+                logger.warn("objectId parameter exceeds maximum string length, truncating.")
+                return@let it.substring(0, MAX_LENGTH_OBJECT_ID)
```

Comment:
This was often surprising to users of the existing library. You added a warning, that's better. I wonder if we should be more aggressive and throw an exception? In the existing library we can't do that anymore because it's widely used and there are events exceeding the limit already. :)

Reply:
If we would throw an exception in production builds, that might not be expected by the engineers and break product features. Is that correct?

I think we should stick with the guideline from desktop of _"don't break product code"_ and keeping it simple for product engineers.
Let's take the best measures to highlight this to users, but don't break product features in release.